sdk/python/kfp/components/_components.py

```diff
@@ -221,15 +226,51 @@ def _generate_input_uri(port_name: str) -> str:
         The URI assigned to this input, will be consistent with the URI where
         the actual content is written after compilation.
     """
-    return os.path.join(
+    # Use hand-crafted path since the behavior of os.path.join varies across
+    # different OSs, making tests difficult.
+    return '{}/{}/{}/{}'.format(
         OUTPUT_DIR_PLACEHOLDER,
-        '{{workflow.uid}}',
+        RUN_ID_PLACEHOLDER,
         '{{{{inputs.parameters.{input}}}}}'.format(
             input=PRODUCER_POD_NAME_PARAMETER.format(port_name)),
         port_name
     )
 
 
+def _generate_output_metadata_path() -> str:
+    """Generates the URI to write the output metadata JSON file."""
+
+    # Use hand-crafted path since the behavior of os.path.join varies across
+    # different OSs, making tests difficult.
+    return '{}/{}/{}'.format(
+        OUTPUT_DIR_PLACEHOLDER,
+        RUN_ID_PLACEHOLDER,
+        OUTPUT_METADATA_JSON
+    )
+
+
+def _generate_input_metadata_path(port_name: str) -> str:
+    """Generates the placeholder for input artifact metadata file."""
+
+    # Return a placeholder for path to input artifact metadata, which will be
+    # rewritten during pipeline compilation.
+    return '{}/{}/{}/{}'.format(
+        OUTPUT_DIR_PLACEHOLDER,
+        RUN_ID_PLACEHOLDER,
+        '{{{{inputs.parameters.{input}}}}}'.format(
+            input=PRODUCER_POD_NAME_PARAMETER.format(port_name)),
+        OUTPUT_METADATA_JSON
+    )
+
+
+def _generate_input_output_name(port_name: str) -> str:
+    """Generates the placeholder for input artifact's output name."""
```

Comment:
Is for passing through an artifact? When would we need this?

Reply:
When we need to pass an artifact produced by a v2 component to a consumer, together with its metadata.

Currently we're writing per-execution output metadata, therefore in order to allow the entrypoint to correctly associate each input artifact with the right field in the metadata JSON file produced by the producer, it needs to know the output name of the artifact when it's being output by the producer.